sway/config/output.c

```diff
@@ -569,3 +465,133 @@ void free_output_config(struct output_config *oc) {
 	free(oc->background_option);
 	free(oc);
 }
+
+static void handle_swaybg_client_destroy(struct wl_listener *listener,
+		void *data) {
+	wl_list_remove(&config->swaybg_client_destroy.link);
+	wl_list_init(&config->swaybg_client_destroy.link);
+	config->swaybg_client = NULL;
+}
+
+static bool set_cloexec(int fd, bool cloexec) {
+	int flags = fcntl(fd, F_GETFD);
+	if (flags == -1) {
+		sway_log_errno(SWAY_ERROR, "fcntl failed");
+		return false;
+	}
+	if (cloexec) {
+		flags = flags | FD_CLOEXEC;
+	} else {
+		flags = flags & ~FD_CLOEXEC;
+	}
+	if (fcntl(fd, F_SETFD, flags) == -1) {
+		sway_log_errno(SWAY_ERROR, "fcntl failed");
+		return false;
+	}
+	return true;
+}
+
+static bool _spawn_swaybg(char *command) {
+	if (config->swaybg_client != NULL) {
+		wl_client_destroy(config->swaybg_client);
+	}
+	int sockets[2];
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockets) != 0) {
+		sway_log_errno(SWAY_ERROR, "socketpair failed");
+		return false;
+	}
+	if (!set_cloexec(sockets[0], true) || !set_cloexec(sockets[1], true)) {
+		return false;
+	}
+
+	config->swaybg_client = wl_client_create(server.wl_display, sockets[0]);
+	if (config->swaybg_client == NULL) {
+		sway_log_errno(SWAY_ERROR, "wl_client_create failed");
+		return false;
+	}
+
+	config->swaybg_client_destroy.notify = handle_swaybg_client_destroy;
+	wl_client_add_destroy_listener(config->swaybg_client,
+		&config->swaybg_client_destroy);
+
+	pid_t pid = fork();
+	if (pid < 0) {
+		sway_log_errno(SWAY_ERROR, "fork failed");
+		return false;
+	} else if (pid == 0) {
+		pid = fork();
+		if (pid < 0) {
+			sway_log_errno(SWAY_ERROR, "fork failed");
+			_exit(EXIT_FAILURE);
+		} else if (pid == 0) {
+			if (!set_cloexec(sockets[1], false)) {
```

Comment:
Wouldn't it make more sense to not set CLOEXEC in the first place?

Reply:
I think it's needed for the double fork for the wayland client socket. _Just as a note, that's currently in master. I just moved the function_

Reply:
We need to set CLOEXEC in the compositor process, but unset it in the child process. So this is really needed.